rasterio/_io.pyx

```diff
@@ -441,19 +456,23 @@ cdef class DatasetReaderBase(DatasetBase):
             raise ValueError("No indexes to read")
 
         # Get the natural shape of the read window, boundless or not.
+
+        # Stub the win_shape.
         win_shape = (len(indexes),)
+
         if window:
-            if boundless:
-                win_shape += (
-                        window[0][1]-window[0][0], window[1][1]-window[1][0])
-            else:
-                w = windows.evaluate(window, self.height, self.width)
-                minr = min(max(w[0][0], 0), self.height)
-                maxr = max(0, min(w[0][1], self.height))
-                minc = min(max(w[1][0], 0), self.width)
-                maxc = max(0, min(w[1][1], self.width))
-                win_shape += (maxr - minr, maxc - minc)
-                window = ((minr, maxr), (minc, maxc))
+
+            if isinstance(window, tuple):
```

Comment:
Done in 361112e. I prefer the tenary clause, we can just delete it when we don't need it.